models/issue_xref.go

```diff
@@ -0,0 +1,309 @@
+	crossReferenceIssueNumericPattern = regexp.MustCompile(`(?:\s|^|\(|\[)([0-9a-zA-Z-_\.]+)/([0-9a-zA-Z-_\.]+)#([0-9]+)(?:\s|$|\)|\]|\.(\s|$))`)
+)
+
+// XRefAction represents the kind of effect a cross reference has once is resolved
+type XRefAction int64
+
+const (
+	// XRefActionNone means the cross-reference is a mention (commit, etc.)
+	XRefActionNone XRefAction = iota // 0
+	// XRefActionCloses means the cross-reference should close an issue if it is resolved
+	XRefActionCloses // 1 - not implemented yet
+	// XRefActionReopens means the cross-reference should reopen an issue if it is resolved
+	XRefActionReopens // 2 - Not implemented yet
+	// XRefActionNeutered means the cross-reference will no longer affect the source
+	XRefActionNeutered // 3
+)
+
+type crossReference struct {
+	Issue  *Issue
+	Action XRefAction
+}
+
+// crossReferencesContext is context to pass along findCrossReference functions
+type crossReferencesContext struct {
+	Type        CommentType
+	Doer        *User
+	OrigIssue   *Issue
+	OrigComment *Comment
+}
+
+func newCrossReference(e *xorm.Session, ctx *crossReferencesContext, xref *crossReference) error {
+	var refCommentID int64
+	if ctx.OrigComment != nil {
+		refCommentID = ctx.OrigComment.ID
+	}
+	_, err := createComment(e, &CreateCommentOptions{
+		Type:         ctx.Type,
+		Doer:         ctx.Doer,
+		Repo:         xref.Issue.Repo,
+		Issue:        xref.Issue,
+		RefRepoID:    ctx.OrigIssue.RepoID,
+		RefIssueID:   ctx.OrigIssue.ID,
+		RefCommentID: refCommentID,
+		RefAction:    xref.Action,
+		RefIsPull:    xref.Issue.IsPull,
+	})
+	return err
+}
+
+func neuterCrossReferences(e Engine, issueID int64, commentID int64) error {
+	active := make([]*Comment, 0, 10)
+	sess := e.Where("`ref_action` IN (?, ?, ?)", XRefActionNone, XRefActionCloses, XRefActionReopens)
+	if issueID != 0 {
+		sess = sess.And("`ref_issue_id` = ?", issueID)
+	}
+	if commentID != 0 {
+		sess = sess.And("`ref_comment_id` = ?", commentID)
+	}
+	err := sess.Find(&active)
+	if err != nil || len(active) == 0 {
+		return err
+	}
+	ids := make([]int64, len(active))
+	for i, c := range active {
+		ids[i] = c.ID
+	}
+	_, err = e.In("id", ids).Cols("`ref_action`").Update(&Comment{RefAction: XRefActionNeutered})
+	return err
+}
+
+// .___
+// |   | ______ ________ __   ____
+// |   |/  ___//  ___/  |  \_/ __ \
+// |   |\___ \ \___ \|  |  /\  ___/
+// |___/____  >____  >____/  \___  >
+//          \/     \/            \/
+//
+
+func (issue *Issue) addCrossReferences(e *xorm.Session, doer *User) error {
+	var commentType CommentType
+	if issue.IsPull {
+		commentType = CommentTypePullRef
+	} else {
+		commentType = CommentTypeIssueRef
+	}
+	ctx := &crossReferencesContext{
+		Type:      commentType,
+		Doer:      doer,
+		OrigIssue: issue,
+	}
+	return issue.findCrossReferences(e, ctx, issue.Title+"\n"+issue.Content)
+}
+
+func (issue *Issue) findCrossReferences(e *xorm.Session, ctx *crossReferencesContext, content string) error {
```

Comment:
Maybe rename it as `createCrossReferences` is better since it will create some comments?

Reply:
You're right.